### PR TITLE
Add several flow features and bugfixes

### PR DESCRIPTION
**Pull request by davidswinegar**
Add coverage checking for Flow, some bugfixes and extra error
handling. Also add two new features - the option to mix sublime
and flow autocomplete, and to omit function parameters for
autocomplete (similar to Nuclide). Add the ability to override
settings on a per-project basis under the "FlowIDE" key, so that
different projects can use different paths for Flow.

Also add an MIT license (discussed over email).

### CODE REVIEW

**Comment by AriaMinaei** [comment]
I just tried this branch. Fixes the two biggest painpoints. I'm very grateful.

I'd say the only thing remaining to bring FlowIDE up to par with Nuclide is the ability to show the flow errors in a small popup box. Because as soon as an error message gets a bit long, I'd have to run `$  flow` inside the terminal to really see what's going on. Afaik this feature is doable because Sublime added a capability that allows packages to do simple HTML rendering a while ago.

**Comment by STRML** [comment]
:+1: Amazing. The coverage output is exactly what I wanted (barely visible but there), and the autocomplete fixes are really great. Really nice work!

**Comment by davidswinegar** [comment]
Hey @tptee - would you be able to take a look at this?

**Comment by tptee** [comment]
Looking now!

File: flow-ide.py [hunk 1]
```diff
@@ -105,16 +137,20 @@ def parse_cli_dependencies(view, **kwargs):
...
     os.close(write)
 
+    # Make sure that we have the default place flow is installed in our $PATH
+    if '/usr/local/bin' not in os.environ['PATH']:
+        os.environ['PATH'] += ':/usr/local/bin'
+
     try:
         output = subprocess.check_output(
-            command, stderr=subprocess.STDOUT, stdin=read
+            command, stderr=subprocess.STDOUT, stdin=read, shell=False
```
Comment (tptee) [inline flow-ide.py:153]: Good call!

File: flow-ide.py [hunk 2]
```diff
@@ -127,7 +163,32 @@ def call_flow_cli(contents, command):
         except:
             os.close(read)
             print(e.output)
+            view.erase_regions('flow_error')
+            view.erase_regions('flow_uncovered')
+            view.set_status('flow_error', 'Unknown Flow error: ' + str(e.output))
             return None
+    except Exception as e:
+        os.close(read)
+        view.erase_regions('flow_error')
```
Comment (tptee) [inline flow-ide.py:172]: This looks copied from above. Can we move this to a utility function like `reset()` or something?

File: flow-ide.py [hunk 3]
```diff
@@ -149,7 +210,7 @@ def run_async(self):
             '--path', deps.filename,
             '--json',
             str(deps.row + 1), str(deps.col + 1)
-        ])
+        ], self.view)
```
Comment (tptee) [inline flow-ide.py:213]: I'd prefer this to be the second argument to `call_flow_cli`

File: flow-ide.py [hunk 4]
```diff
@@ -260,6 +328,11 @@ def on_query_completions_async(self, view, prefix, locations):
 
     @wait_for_load
     def on_selection_modified_async(self, view):
+        flow_settings = find_flow_settings(view.window().project_data())
+        # print('Selection modified, delaying %d' % flow_settings.get('flow_debounce_ms'))
```
Comment (tptee) [inline flow-ide.py:332]: Nit: commented code

File: flow-ide.py [hunk 5]
```diff
@@ -268,71 +341,107 @@ def on_selection_modified_async(self, view):
...
             deps.filename
-        ])
+        ], view)
 
         if result:
             if result['passed']:
                 view.erase_regions('flow_error')
                 view.set_status('flow_error', 'Flow: no errors')
-                return
+            else:
```
Comment (tptee) [inline flow-ide.py:366]: Can we do an early `return` before this so that all the coverage code isn't indented so far?

**Review by tptee — APPROVED** [review]
Works wonders! I have some stylistic comments, but I think those are mainly the result of this single file getting too big. I'm going to go ahead and merge, then split out some of `flow-ide.py`.

Thanks for your contribution!! 🎉

**Comment by tptee** [comment]
Sorry for the wait, I went down a refactoring rabbit hole 😬 `v1.2.0` is published and Package Control should pick up the update within the next 30 mins.

**Pull request merged**